modules/core/src/main/scala/io/chrisdavenport/mules/DispatchOneCache.scala

```diff
@@ -250,71 +253,148 @@ object DispatchOneCache {
       defaultExpiration
     )
   }
-
   def ofMapRef[F[_]: Concurrent: Clock, K, V](
     mr: MapRef[F, K, Option[DispatchOneCacheItem[F, V]]],
-    defaultExpiration: Option[TimeSpec]
+    defaultExpiration: Option[TimeSpec],
+    purgeExpiredEntries: Option[Long => F[List[K]]]
   ): DispatchOneCache[F, K, V] = {
     new DispatchOneCache[F, K, V](
         mr,
-        None,
+        purgeExpiredEntries,
         defaultExpiration
       )
   }
 
+  private[mules] def isExpired[F[_], A](checkAgainst: Long, cacheItem: DispatchOneCacheItem[F, A]): Boolean = {
+    cacheItem.itemExpiration match{
+      case Some(e) if e.nanos < checkAgainst => true
+      case _ => false
+    }
+  }
+}
 
-  private object SingleRef {
-
-    def purgeExpiredEntries[F[_], K, V](ref: Ref[F, Map[K, DispatchOneCacheItem[F, V]]])(now: Long): F[List[K]] = {
-      ref.modify(
-        m => {
-          val l = scala.collection.mutable.ListBuffer.empty[K]
-          m.foreach{ case (k, item) =>
-            if (isExpired(now, item)) {
-              l.+=(k)
-            }
+private[mules] object SingleRef {
+  def purgeExpiredEntries[F[_], K, V](ref: Ref[F, Map[K, V]], isExpired: (Long, V) => Boolean)(now: Long): F[List[K]] = {
+    ref.modify(
+      m => {
+        val l = scala.collection.mutable.ListBuffer.empty[K]
+        m.foreach { case (k, item) =>
+          if (isExpired(now, item)) {
+            l.+=(k)
           }
-          val remove = l.result()
-          val finalMap = m -- remove
-          (finalMap, remove)
         }
+        val remove = l.result()
+        val finalMap = m -- remove
+        (finalMap, remove)
+      }
+    )
+  }
+}
+
+private[mules] case class PurgeableMapRef[F[_],K,V](mapRef: MapRef[F,K,V], purgeExpiredEntries: Long => F[List[K]])
+
+private[mules] object PurgeableMapRef {
+  def ofShardedImmutableMap[F[_]: Concurrent, K, V](
+    shardCount: Int,
+    isExpired: (Long, V) => Boolean
+  ): F[PurgeableMapRef[F, K, Option[V]]] = {
+    assert(shardCount >= 1, "MapRef.sharded should have at least 1 shard")
+    val shards: F[List[Ref[F, Map[K, V]]]] =   List.fill(shardCount)(())
+      .traverse(_ => Concurrent[F].ref[Map[K, V]](Map.empty))
+
+    def purgeExpiredEntries(shards:List[Ref[F, Map[K, V]]])(now: Long) = shards.parFlatTraverse(SingleRef.purgeExpiredEntries(_, isExpired)(now))
+
+    shards.map{ s =>
+      PurgeableMapRef(
+        fromSeqRefs(s),
+        purgeExpiredEntries(s)
       )
     }
   }
 
-  private def isExpired[F[_], A](checkAgainst: Long, cacheItem: DispatchOneCacheItem[F, A]): Boolean = {
-    cacheItem.itemExpiration match{
-      case Some(e) if e.nanos < checkAgainst => true
-      case _ => false
+  def ofConcurrentHashMap[F[_]: Concurrent, K, V](
+    initialCapacity: Int = 16,
+    loadFactor: Float = 0.75f,
+    concurrencyLevel: Int = 16,
+    isExpired: (Long, V) => Boolean
+  ): F[PurgeableMapRef[F, K, Option[V]]] = Concurrent[F].unit.map{ _ => //replaced Sync[F].delay.  Needed?
```

Comment:
Depends how much we are breaking the rules.

Reply:
Hmm.  Not sure how to interpret that one.  It was basically the same construction is the DispatchOneCache and MemoryCache files, but changing that one would have required increasing the types all the way up the chain, so rather than bring those types up, I brought this one down (to paraphrase Alice's Restaurant).  I'm not clear if this has an implications for what goes on under the hood here.